nimbus/src/enrollment.rs

```diff
@@ -311,55 +758,820 @@ mod tests {
+                ..
+            }
+        ));
+        assert_eq!(events.len(), 1);
+        assert!(matches!(
+            events[0].change,
+            EnrollmentChangeEventType::Enrollment
+        ));
+        Ok(())
+    }
+
+    #[test]
+    fn test_enrollment_explicit_opt_in_branch_unknown() {
+        let exp = get_test_experiments()[0].clone();
+        let mut events = vec![];
+        let res = ExperimentEnrollment::from_explicit_opt_in(&exp, "bobo", &mut events);
+        assert!(res.is_err());
+    }
+
+    #[test]
+    fn test_enrollment_enrolled_explicit_opt_out() -> Result<()> {
+        let exp = get_test_experiments()[0].clone();
+        let mut events = vec![];
+        let enrollment_id = Uuid::new_v4();
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug,
+            status: EnrollmentStatus::Enrolled {
+                enrollment_id,
+                branch: "control".to_owned(),
+                reason: EnrolledReason::Qualified,
+            },
+        };
+        let enrollment = existing_enrollment.on_explicit_opt_out(&mut events)?;
+        if let EnrollmentStatus::Disqualified {
+            enrollment_id: new_enrollment_id,
+            ..
+        } = enrollment.status
+        {
+            assert_eq!(new_enrollment_id, enrollment_id);
+        } else {
+            panic!("Wrong variant!");
+        }
+        assert_eq!(events.len(), 1);
+        assert!(matches!(
+            events[0].change,
+            EnrollmentChangeEventType::Disqualification
+        ));
+        Ok(())
+    }
+
+    #[test]
+    fn test_enrollment_not_enrolled_explicit_opt_out() -> Result<()> {
+        let exp = get_test_experiments()[0].clone();
+        let mut events = vec![];
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug,
+            status: EnrollmentStatus::NotEnrolled {
+                reason: NotEnrolledReason::NotTargeted,
+            },
+        };
+        let enrollment = existing_enrollment.on_explicit_opt_out(&mut events)?;
+        assert!(
+            matches!(enrollment.status, EnrollmentStatus::NotEnrolled { reason: NotEnrolledReason::OptOut, .. })
+        );
+        assert!(events.is_empty());
+        Ok(())
+    }
+
+    #[test]
+    fn test_enrollment_disqualified_explicit_opt_out() -> Result<()> {
+        let exp = get_test_experiments()[0].clone();
+        let mut events = vec![];
+        let existing_enrollment = ExperimentEnrollment {
+            slug: exp.slug,
+            status: EnrollmentStatus::Disqualified {
+                enrollment_id: Uuid::new_v4(),
+                branch: "control".to_owned(),
+                reason: DisqualifiedReason::NotTargeted,
+            },
+        };
+        let enrollment = existing_enrollment.on_explicit_opt_out(&mut events)?;
+        assert_eq!(enrollment, existing_enrollment);
+        assert!(events.is_empty());
+        Ok(())
+    }
```

Comment:
A *fantastic* suite of tests, thank you!

For a future iteration, I think it would be neat if we could make these a little more declarative, so that someone like Kate or Tim could review them for correctness without needing to deeply understand the details here. But that's definitely for a followup, not this PR!